consoles/sshVirtsh.pm

```diff
@@ -325,111 +325,169 @@ sub add_interface {
     return;
 }
 
+sub to_byte {
+    my $sp              = shift;
+    my $size_multiplier = {
+        b => 1,
+        k => 1024, K => 1024,
+        M => 1024 * 1024,
+        G => 1024 * 1024 * 1024,
+        T => 1024 * 1024 * 1024 * 1024,
+    };
+
+    #in case of invalid format return error
+    if (!defined($sp) || $sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
+        return undef;
+    }
+
+    return $+{num} * $size_multiplier->{$+{unit} // 'b'};
+}
+
+sub gen_kvm_img {
+    my ($self, %args) = @_;
+    my $byte_size = to_byte($args{size});
+    die('Missing argument size or invalid format') unless (defined($byte_size));
+
+    push(my @cmd, 'qemu-img', 'create');
+    push(@cmd,    '-f',       'qcow2');
+    my $size = $args{size};
+    if ($args{backingfile}) {
+        push(@cmd, '-b', $args{backingfile});
+        # expected value in Bytes
+        my (undef, $json) = $self->run_cmd("qemu-img info --output=json $args{backingfile}", wantarray => 1);
+        my $image_vsize = decode_json($json)->{'virtual-size'};
+        if ($byte_size <= $image_vsize) {
+            $size = $image_vsize;
+        }
+    }
+    push(@cmd, $args{new_image_file});
+    push(@cmd, $size);
+    bmwqemu::diag("gen_kvm_img: Creating image file...");
+    my $tries = 5;
+    # Avoid qemu-img's failure to get a write lock to be the reason for a job to fail
+    my ($ret, $stdout, $stderr);
+    for (1 .. $tries) {
+        ($ret, $stdout, $stderr) = $self->run_cmd(join(" ", @cmd), wantarray => 1);
+        if ($stderr) {
+            bmwqemu::diag("gen_kvm_img failed: $stderr, try $_ out of $tries");
+            sleep 5;
+        }
+        last unless $ret;
+    }
+    die "Too many attempts to format HDD" if $stderr;
+    bmwqemu::diag("gen_kvm_img: Finished creating image file...");
+}
+
+=head2 download_image_file
+
+Internal method to download asset files using rsync. Requires absolute path to source file
+and destination directory. If successful, returns absolute path to downloaded file
+
+    $self->download_image_file('/var/lib/openqa/share/factory/hdd/SUSE-MicroOS.s390x-5.1.0-Default-kvm-Build10.9.qcow2', '/var/lib/libvirt/images/');
+
+=cut
+
+sub download_image_file {
+    my ($self, $src, $dest) = @_;
+    $dest .= basename($src);
+    die "Downloading $src failed!" if $self->run_cmd("rsync -av $src $dest");
+    return $dest;
+}
+
+sub decompress_cmd {
+    return sprintf(q(xz --decompress --keep --verbose --no-warn --quiet '%s'), shift);
+}
+
 sub add_disk {
```

Comment:
The method is quite big and complex. Could you please split the method in a separate commit before changing it? If you are unsure how to do that I could try myself to suggest a change.

Reply:
Yes, please, it would be nice if you can try to split the original method. Thanks

Reply:
I did so in https://github.com/os-autoinst/os-autoinst/pull/1752 . I am sure this will cause multiple conflicts for you and I have only tested it with t/22-svirt.t so feel free to incorporate that in anyways helpful to you and please test accordingly.